### PR TITLE
API to get Expr's type and nullability without a `DFSchema`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/1725

 # Rationale for this change
I am trying to create a low cost (low copy) way for `Expr` simplification.  More details on https://github.com/apache/arrow-datafusion/issues/1725

# What changes are included in this PR?
1. Add a trait that can provide the needed information for schema quries to Expr
1. Change the functions on `Expr` that take a &DFSchema such as `Expr::nullable` to be nullable
3. `impl` this new trait for `DFSchema` so that existing callsites work

# Are there any user-facing changes?
1. It is now possible to get an Expr's type without constructing a `DFSchema`; Existing code should continue to work

### CODE REVIEW

**Comment by alamb** [comment]
I am happy to improve / change this API -- if anyone has thoughts or suggestions please let me know

File: datafusion/src/logical_plan/expr.rs [hunk 1]
```diff
@@ -392,20 +392,54 @@ impl PartialOrd for Expr {
...
+    }
+
+    fn data_type(&self, col: &Column) -> Result<&DataType> {
+        Ok(self.field_from_column(col)?.data_type())
+    }
+}
+
 impl Expr {
-    /// Returns the [arrow::datatypes::DataType] of the expression based on [arrow::datatypes::Schema].
+    /// Returns the [arrow::datatypes::DataType] of the expression based on [DFSchema].
```
Comment (houqp) [inline datafusion/src/logical_plan/expr.rs:430]: I don't think this comment is accurate anymore since this method is not restrict to `DFSchema` anymore.

**Review by houqp — APPROVED** [review]
nice idea to extend the interface :+1:

File: datafusion/src/logical_plan/expr.rs (hunk 1 above)
Comment (alamb) [reply datafusion/src/logical_plan/expr.rs:430]: good point -- will fix

**Pull request merged**